test/core/ticktree/TickTreeMarketOrder.t.sol

```diff
@@ -0,0 +1,243 @@
+// SPDX-License-Identifier:  AGPL-3.0
+
+pragma solidity ^0.8.18;
+
+import "./TickTreeTest.t.sol";
+import "mgv_src/MgvLib.sol";
+import "mgv_lib/Debug.sol";
+
+// Tests of Mangrove.marketOrder's interaction with the tick tree.
+//
+// The tests use the following pattern:
+// 1. we establish a Mangrove tick tree where there may be offers at:
+//   - a lower tick
+//   - a middle tick
+//   - a higher tick
+// 2. we take a snapshot of Mangrove's tick tree
+// 3. we run a market order in Mangrove and remove the corresponding offers in the snapshot tick tree
+// 4. we check that Mangrove's tick tree matches the test tick tree.
+//
+// The scenarios we want to test are:
+// - lower tick list
+//   - tick is MIN, MAX, in same {leaf, level0, level1, level2} as middle tick
+//     - if feasible, given middle tick
+//   - list:
+//     1. is empty
+//     2. is fully taken
+//     3. is partially taken
+//     4. is not taken
+// - middle tick (in lower tick cases 1. and 2.)
+//   - tick is MIN, MAX, min&max&mid {leaf, level0, level1, level2}
+//   - list:
+//     1. is empty
+//     2. is fully taken
+//     3. is partially taken
+//     4. is not taken
+// - higher tick list (in middle tick cases 1. and 2.)
+//   - tick is MIN, MAX, in same {leaf, level0, level1, level2} as middle tick
+//     - if feasible, given middle tick
+//   - list:
+//     1. is empty
+//     2. is fully taken
+//     3. is partially taken
+//     4. is not taken
+//
+// We do not test failing offers or partial fills specifically,
+// as they are not handled specially wrt the tick tree.
+contract TickTreeMarketOrderTest is TickTreeTest {
+  function setUp() public override {
+    super.setUp();
+
+    // Check that the tick tree is consistent after set up
+    assertMgvTickTreeIsConsistent();
+  }
+
+  // Tick list               size  offersToTake
+  // 1. is empty                0  0
+  // 2. is fully taken          1  1
+  // 3. is partially taken      2  1
+  // 4. is not taken            2  0
+  struct TickListScenario {
+    Tick tick;
+    uint size;
+    uint offersToTake;
+  }
+
+  struct MarketOrderScenario {
+    TickListScenario lowerTick;
+    TickListScenario middleTick; // ignored if lowerTickListScenario.size == 2 (scenario 3. and 4.)
+    TickListScenario higherTick; // ignored if {lower,middle}TickListScenario.size == 2 (scenario 3. and 4.)
+  }
+
+  // (list size, offers to take)
+  uint[2][] tickListScenarios = [[0, 0], [1, 1], [2, 1], [2, 0]];
```

Comment:
so we never take 2 offers?

Reply:
Indeed; maybe we should?

Maybe a single test that ensures we can take two offers would be sufficient to test the tick list updates in market orders?

Reply:
Tick list scenarios have been changed such that fully taken lists have length 2: 150c33c1bb456ff1f982118d155bdf2c65917f4a